### ISSUE TITLE
Energy is NaN error while using periodicdistance() function

### ISSUE THREAD

**Issue by dr-nate** [pandegroup/openmm#1191]
Hey all,

I've been trying to use the new periodicdistance() function to restrain an atom for NPT equilibration, but at the first energy check it throws an error:

```
#"Progress (%)"	"Step"	"Time (ps)"	"Potential Energy (kJ/mole)"	"Kinetic Energy (kJ/mole)"	"Total Energy (kJ/mole)"	"Temperature (K)"	"Box Volume (nm^3)"	"Speed (ns/day)"	"Time Remaining"
Traceback (most recent call last):
  File "openmm-equil.py", line 131, in <module>
    simulation.step(numsteps)
  File "/Users/stanlen1/anaconda/lib/python3.4/site-packages/simtk/openmm/app/simulation.py", line 96, in step
    self._simulate(endStep=self.currentStep+steps)
  File "/Users/stanlen1/anaconda/lib/python3.4/site-packages/simtk/openmm/app/simulation.py", line 173, in _simulate
    reporter.report(self, state)
  File "/Users/stanlen1/anaconda/lib/python3.4/site-packages/simtk/openmm/app/statedatareporter.py", line 162, in report
    self._checkForErrors(simulation, state)
  File "/Users/stanlen1/anaconda/lib/python3.4/site-packages/simtk/openmm/app/statedatareporter.py", line 309, in _checkForErrors
    raise ValueError('Energy is NaN')
ValueError: Energy is NaN
```
Looking deeper into the issue by writing out parameters for that atom every frame, it seems that the position and force on the particle (a single atom) are NaN from the first step of the simulation. Here is the code for applying the restraint:

``` 
# Add positional restraints to ion
force = mm.CustomExternalForce("k*periodicdistance(x, y, z, x0, y0, z0)^2")
force.addGlobalParameter("k", 5.0*u.kilocalories_per_mole/u.angstroms**2)
force.addPerParticleParameter("x0")
force.addPerParticleParameter("y0")
force.addPerParticleParameter("z0")
for i, atom_crd in enumerate(parm.positions):
    if parm.atoms[i].name in ('K+'):
        print(atom_crd.value_in_unit(u.nanometers))
        force.addParticle(i, atom_crd.value_in_unit(u.nanometers))
system.addForce(force)
```
If I swap out that first line with the one below, the simulation can usually complete without issue. 

```
force = mm.CustomExternalForce("k*((x-x0)^2+(y-y0)^2+(z-z0)^2)")
```

I am using a personally compiled version of the code from commit b38618a72834d12207cf14de06153f43c2c32318. However, I have similar issues using the openmm-dev branch from anaconda as well (but there I get 'Particle position is NaN' error because of the new check that was recently added).

I've tar-ed up a folder with all the necessary files from the test system I've been using to test this. 

https://dl.dropboxusercontent.com/u/10524044/ionwaterbox.tgz

Is there an issue with the code or am I doing something stupid? I don't know the openmm codebase well enough yet to know where things are going wrong.

**Comment by rmcgibbo** [comment]
I can confirm this. Here's a simple example that reproduces the problem

```
import numpy as np
import simtk.openmm as mm

system = mm.System()
system.addParticle(1)

force = mm.CustomExternalForce("k*periodicdistance(x, y, z, x0, y0, z0)^2")
force.addGlobalParameter("k", 1.0)
force.addPerParticleParameter("x0")
force.addPerParticleParameter("y0")
force.addPerParticleParameter("z0")
force.addParticle(0, [1., 1., 1.])
system.addForce(force)

integrator = mm.LangevinIntegrator(300, 1.0, 0.002)
platform = mm.Platform.getPlatformByName('CPU')
context = mm.Context(system, integrator, platform)

context.setPositions([[1., 1., 1.]])
system.setDefaultPeriodicBoxVectors([2,0,0], [0,2,0], [0,0,2])
context.setPeriodicBoxVectors([2,0,0], [0,2,0], [0,0,2])

integrator.step(1)
pos = context.getState(getPositions=True).getPositions(asNumpy=True)
print(pos)
```

**Comment by rmcgibbo** [comment]
A fix is in #1192. The bug is pretty clear if you look [here](https://github.com/pandegroup/openmm/blob/213fbc0/platforms/reference/src/ReferenceKernels.cpp#L1490-L1493). I'm sure there's probably a more efficient way to do the check on CUDA/OpenCL though.

**Comment by peastman** [comment]
Fixed by #1192.
